Umee-Bots/High-Utilization-Ratio/src/agent.ts

```diff
@@ -0,0 +1,150 @@
+  ethers,
+  Finding,
+  getEthersProvider,
+  HandleBlock,
+  HandleTransaction,
+  Initialize,
+  TransactionEvent,
+} from "forta-agent";
+import CONFIG from "./agent.config";
+import { GET_RESERVES_LIST_ABI, GET_RESERVE_DATA_ABI, RESERVE_INITIALIZED_ABI } from "./constants";
+import { MulticallProvider } from "./multicall";
+import {
+  AgentConfig,
+  arrayChunks,
+  createAbsoluteThresholdFinding,
+  createPercentageThresholdFinding,
+  createReserveData,
+  ReserveBalances,
+  ReserveData,
+  usageRatio,
+} from "./utils";
+
+BigNumber.set({ DECIMAL_PLACES: 18 });
+
+const reserveData: ReserveData[] = [];
+
+// testing
+export const resetReserveData = () => (reserveData.length = 0);
+export const getReserveData = () => reserveData;
+export const setReserveData = (newReserveData: ReserveData[]) => {
+  resetReserveData();
+  newReserveData.forEach((el) => reserveData.push(el));
+};
+
+let multicallProvider: MulticallProvider;
+
+export const provideInitialize = (provider: ethers.providers.Provider, config: AgentConfig): Initialize => {
+  return async () => {
+    const LendingPool = new ethers.Contract(
+      config.lendingPoolAddress,
+      [GET_RESERVES_LIST_ABI, GET_RESERVE_DATA_ABI],
+      provider
+    );
+
+    const reserveAddresses: string[] = await LendingPool.getReservesList();
+    const reserves = await Promise.all(reserveAddresses.map((el) => LendingPool.getReserveData(el)));
+
+    reserves.forEach((data, idx) => {
+      reserveData.push(
+        createReserveData(
+          reserveAddresses[idx],
+          data.uTokenAddress,
+          data.stableDebtTokenAddress,
+          data.variableDebtTokenAddress
+        )
+      );
+    });
+
+    multicallProvider = new MulticallProvider(provider);
+    await multicallProvider.init();
+  };
+};
+
+export const provideHandleTransaction = (config: AgentConfig): HandleTransaction => {
+  return async (txEvent: TransactionEvent): Promise<Finding[]> => {
+    const reserveInits = txEvent.filterLog(RESERVE_INITIALIZED_ABI, config.lendingPoolConfiguratorAddress);
+
+    reserveInits.forEach((reserveInit) => {
+      reserveData.push(
+        createReserveData(
+          reserveInit.args.asset,
+          reserveInit.args.uToken,
+          reserveInit.args.stableDebtToken,
+          reserveInit.args.variableDebtToken
+        )
+      );
+    });
+
+    return [];
+  };
+};
+
+export const provideHandleBlock = (config: AgentConfig): HandleBlock => {
+  return async (blockEvent: BlockEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+
+    // make one multicall for each chunk of 10 reserves
+    const multicalls = arrayChunks(reserveData, 10).map((chunk) => {
+      return multicallProvider.all(
+        chunk.flatMap((reserve) => [
+          reserve.asset.balanceOf(reserve.uTokenAddress),
+          reserve.stableDebtToken.totalSupply(),
+          reserve.variableDebtToken.totalSupply(),
+        ]),
+        blockEvent.blockNumber
+      );
+    });
```

Comment:
Yes! That's because providers do have "gas limits" for calls, so it's better to try and make sure the call won't fail because of that.

By the way, I did a stress test on both this bot and UM01 with the Alchemy providers, and they both seemed to work fine at max chunk capacity.